src/main/java/net/neoforged/neoforge/capabilities/BlockCapabilityCache.java

```diff
@@ -0,0 +1,120 @@
+    /**
+     * Creates a new cache instance and register it to the level.
+     *
+     * @param capability the capability
+     * @param level      the level
+     * @param pos        the position
+     * @param context    extra context for the query
+     */
+    public static <T, C> BlockCapabilityCache<T, C> create(BlockCapability<T, C> capability, ServerLevel level, BlockPos pos, C context) {
+        return create(capability, level, pos, context, () -> true, () -> {});
+    }
+
+    /**
+     * Creates a new cache instance with an invalidation listener, and register it to the level.
+     *
+     * @param capability           the capability
+     * @param level                the level
+     * @param pos                  the position
+     * @param context              extra context for the query
+     * @param isValid              a function to check if the listener still wants to receive notifications.
+     *                             A typical example would be {@code () -> !this.isRemoved() } for a block entity
+     *                             that should not receive invalidation notifications anymore once it is removed.
+     * @param invalidationListener the invalidation listener. Will be called whenever the capability of the cache might have changed.
+     */
+    public static <T, C> BlockCapabilityCache<T, C> create(BlockCapability<T, C> capability, ServerLevel level, BlockPos pos, C context, BooleanSupplier isValid, Runnable invalidationListener) {
+        Objects.requireNonNull(capability);
+        Objects.requireNonNull(isValid);
+        Objects.requireNonNull(invalidationListener);
+        pos = pos.immutable();
+
+        var cache = new BlockCapabilityCache<>(capability, level, pos, context, isValid, invalidationListener);
+        level.registerCapabilityListener(pos, cache.listener);
+        return cache;
+    }
+
+    private final BlockCapability<T, C> capability;
+    private final ServerLevel level;
+    private final BlockPos pos;
+    private final C context;
+
+    private boolean cacheValid = false;
+    @Nullable
+    private T cachedCap = null;
+
+    private boolean canQuery = true;
+    private final ICapabilityInvalidationListener listener;
+
+    private BlockCapabilityCache(BlockCapability<T, C> capability, ServerLevel level, BlockPos pos, C context, BooleanSupplier isValid, Runnable invalidationListener) {
+        this.capability = capability;
+        this.level = level;
+        this.pos = pos;
+        this.context = context;
+
+        this.listener = () -> {
+            if (!cacheValid)
+                return true; // cache was already invalidated...
+
+            canQuery = false;
+            cacheValid = false;
+
+            if (isValid.getAsBoolean()) {
+                invalidationListener.run();
+                canQuery = true;
+                return true;
+            } else {
+                return false;
+            }
+        };
```

Comment:
This listener could use a few more comments as I will be honest when I was first trying to figure out how the implementation worked it was rather confusing. Key parts that should semi be somehow specified:
- Before `getCapability` has been called the listener will just NO-OP
- Remove the cached value and prevent querying during validation checking and invalidation
- Checks if it is still valid, if not it will get marked for removal and this cache will be permanently disabled. Otherwise it runs the invalidation listener and then marks it as queryable again

Secondary question that I think led to why this implementation was confusing and felt like the `if (!cacheValid)` should `return false`... Should that return statement instead run the `isValid` check?